### DISCUSSION TITLE
Is This Possible: Lazy-Loading Remote Modules Dynamically

### DISCUSSION THREAD

**Discussion by BenByrd92** [webpack/webpack#16952] [Q&A]
Hello,

I am trying to make the Module Federation logic for my project totally dynamic, so that we can add a line of 'application metadata' to our SSO DB and load a new remote (built from an MFE boilerplate), without changing any code on the Front End.

So far, I've been almost completely successful. Using a list of remote modules and their URLs I am able to dynamically load all of their remoteEntry.js files. I've also been able to generate routes and types for their <App> modules, but I am down to one final problem, and I'm not sure it is possible now.

I have a component, RemoteApplicationLoader:
```
import React, { lazy, Suspense } from 'react';
import { CircularProgress } from '@inmarket/inmarkit';

const RemoteApplicationLoader = (props: { name: string }) => {
  const { name } = props;
  const path = `${name}/App`;
  const RemoteApplication = lazy(() => import(`${path}`));

  return (
    <Suspense fallback={<CircularProgress />}>
      <RemoteApplication basename={name} />
    </Suspense>
  );
};

export default RemoteApplicationLoader;
```

`RemoteApplicationLoader`s are generated as elements for the Routes, as follows, where `moduleName` can be, for example, "geolink":
```
<Route
  key={`${productGroup.name}-route-${featureIndex}`}
  path={`${moduleName}/*`}
  element={<RemoteApplicationLoader name={moduleName}/>}
/>
 ```

Again, all of this works fine, up to the point where I have to lazy-load the module dynamically. This line does not work:
`const RemoteApplication = lazy(() => import(`${path}`));`
While this line DOES work: 
`const RemoteApplication = lazy(() => import('geolink/App'));`

The ${path} variable absolutely evaluates to the same result that lazy() loads successfully, so this leads me to believe that lazy() will not let me load Webpack Module Federation modules dynamically. Is this conclusion correct? If so, is what I'm trying to do actually possible at all? If not, please let me know what I am doing wrong!

It may be worth noting that if I switch the lazy load line to `import(path)` instead of `import(`${path}`))`, the remoteEntry.js files no longer load. So, it seems like it's close to working, but not quite.

The lazy-load also works if the variable evaluates to a relative path within the same project. It only fails once it evaluates to an MFE module.

 Thanks again in advance!

EDIT: forgot to add the error from the console:
`remoteEntry.js:2 Error: Cannot find module 'geolink/App'
    at  lazy ^.*$ namespace object:656:1
u`

**Comment by alexander-akait** [comment]
Please use https://github.com/webpack/webpack/discussions for questions, thank you

```js
const RemoteApplication = lazy(() => import(`${path}`));
```

is not analizable by webpack

We have an example here - https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers

**Comment by BenByrd92** [comment]
This looks like exactly what I need! Thanks so much @alexander-akait - will give it a shot!

**Comment by BenByrd92** [comment]
Hello @alexander-akait  - thanks for sharing the example. This is definitely the right track, but do you know if I'll be able to generate something like a `<Route element={FederatedModule} path={module.name} />` for each remote I want to use? I know I won't be able to use that `useFederatedComponent` hook inside of a loop. Just want to make sure what I want to do is possible before I continue trying to adapt the solution! Thanks again.

**Reply by alexander-akait** [reply]
You can generate manifest plugin, there are different approaches to do it:
1. If you have file based logic, you can just glob them, create `manifest.json` and load it before, just create s simple plugin (anyway you can do it even without webpack)
2. If you need more complex logic you can use `getStats().toJson()` and get infromation from there
3. If you need even more complex, let us say - a plugin system, you will need own loader to inject plugin logic and plugin to collect infromation from modules.

We don't have special built-in solution, so make your own

**Reply by BenByrd92** [reply]
Thanks @alexander-akait. You're probably right, but I did still want to try one more way that I might accomplish this using your custom `useFederatedComponent` hook. I was having hopes that the following solution might work...

What I tried was creating a FederatedModuleWrapper that would accept props `{ appModule: string, appScope: string, appUrl: string }` and then pass those props to your function.

```
const FederatedModuleWrapper = (appUrl, appScope, appModule) => {
  const [{ url, scope, module }] = useState({
    module: appModule,
    scope: appScope,
    url: appUrl,
  });
  ......
  ```
  and then in App.tsx, generating routes as follows (with hardcoded arguments as an example):
  ```
              <Route
                    key={`${productGroup.name}-route-${featureIndex}`}
                    path={`${moduleName}/*`}
                    element={
                      <FederatedModuleWrapper
                        appModule="./App"
                        appScope="geolink"
                        appUrl="https://geolink.inmarket.com/remoteEntry.js"
                      />
                    }
                  />
```
 This does work if I setup the initial state with hardcoded strings, but when I switch it to props it no longer loads the module.

I know you essentially already answered my question, but in distant hopes... was I onto anything here or is a custom solution still definitely needed? Thanks again so much for your time and expertise,

Ben

**Reply by BenByrd92** [reply]
Following up here with some more findings. I really feel like this latest implementation should work! When the FederatedComponent is setup, all of the variable names evaluate properly, and it even loads the remoteEntry.js, so it must be close...



![Screenshot 2023-04-10 at 6 18 03 PM](https://user-images.githubusercontent.com/91070409/231009834-408ed07c-2420-4a59-aef1-9a1f6d22defc.png)



![Screenshot 2023-04-10 at 6 18 23 PM](https://user-images.githubusercontent.com/91070409/231009835-aed53863-0913-47e8-a13e-1b581d829e30.png)

**Reply by alexander-akait** [reply]
Yes, it's okay to look for the right logic for your application, webpack doesn't know anything about your application and how it works, but provides the ability to write a plugin to get the values that you need
